### PR TITLE
Jobs dependent on finished jobs stay in held state

### PR DESCRIPTION
**Pull request by arungrover**
#### Describe Bug or Feature
If a job is dependent on a job that has already finished by the time dependent job enters the system, it stays in "H" state for the rest of its life.
This change addresses this issue and releases the dependency if it meets the dependency criteria.

#### Describe Your Change
When job history is enabled and PBS server is able to find the finished job, it checks that the exit status of the finished job satisfies the dependency criteria, if it does then no dependency is added on the submitted job and it stays in 'Q' state. If the criteria is not met then the submitted job is simply rejected (the same way it is rejected when job history is not enabled and PBS is unable to find the parent job).

#### Link to Design Doc
NA

#### Attach Test and Valgrind Logs/Output
[test.txt](https://github.com/PBSPro/pbspro/files/4417711/test.txt)

### CODE REVIEW

File: src/server/req_register.c [hunk 1]
```diff
@@ -130,6 +130,8 @@ extern char *msg_regrej;
 extern char *msg_job_moved;
 extern char *msg_depend_runone;
 extern char  log_buffer[];
+extern char *msg_histdepend;
+extern char *msg_historyjobid;
```
Comment (vstumpf) [inline src/server/req_register.c:134]: do you mind moving these above the log_buffer[]? Keep the messages together

File: src/server/req_register.c [hunk 2]
```diff
@@ -294,64 +305,79 @@ req_register(struct batch_request *preq)
...
+							/* release back to child and continue with	*/
+							/* registration process 			*/
+							ptask = set_task(WORK_Immed, 0, post_run_depend,
+								(void *)pjob);
+							if (ptask)
+								append_link(&pjob->ji_svrtask,
+									&ptask->wt_linkobj, ptask);
+						}
+					} else
+						rc = PBSE_HISTJOBID;
```
Comment (vstumpf) [inline src/server/req_register.c:320]: Shouldn't you break after setting rc?

File: src/server/req_register.c [hunk 3]
```diff
@@ -294,64 +305,79 @@ req_register(struct batch_request *preq)
...
-						/* ok owner, see if job has "on" */
-						pdep = find_depend(JOB_DEPEND_TYPE_ON, pattr);
-						if (pdep == 0) {
-							/* on "on", see if child already registered */
-							revtype = type ^ (JOB_DEPEND_TYPE_BEFORESTART -
-								JOB_DEPEND_TYPE_AFTERSTART);
-							pdep = find_depend(revtype, pattr);
+					/* There is no need to put before dependency on a finished job */
+					if (is_finished == FALSE) {
+						/*
```
Comment (vstumpf) [inline src/server/req_register.c:343]: Can you change this with
```C
if (is_finished == TRUE) {
    rc = PBSE_HISTDEPEND;
    break;
}
```
Then you don't have to mess with the indentation and the git blame.

File: src/server/req_register.c (hunk 2 above)
Comment (vstumpf) [reply src/server/req_register.c:320]: You can also do the if (is_finished == TRUE){ rc = PBSE; break; } here

File: src/server/req_register.c [hunk 4]
```diff
@@ -294,64 +305,79 @@ req_register(struct batch_request *preq)
...
-							(void *)pjob);
-						if (ptask)
-							append_link(&pjob->ji_svrtask,
-								&ptask->wt_linkobj, ptask);
-					}
+					if (is_finished == FALSE) {
+						if (pjob->ji_qs.ji_substate >= JOB_SUBSTATE_RUNNING) {
+							/* job already running, setup task to send	*/
+							/* release back to child and continue with	*/
+							/* registration process 			*/
```
[the remark this answers is not in the document]
Comment (arungrover) [reply src/server/req_register.c:312]: Oh, this was an old existing comment. I'll change it.

File: src/server/req_register.c (hunk 2 above)
Comment (arungrover) [reply src/server/req_register.c:320]: Good catch :)

**Comment by arungrover** [comment]
[test.txt](https://github.com/PBSPro/pbspro/files/4430457/test.txt)

File: src/server/req_register.c [hunk 5]
```diff
@@ -327,23 +356,23 @@ req_register(struct batch_request *preq)
 						rc = PBSE_PERM;		/* not same user */
 					} else {
 						/* ok owner, see if job has "on" */
-						pdep = find_depend(JOB_DEPEND_TYPE_ON, pattr);
+							pdep = find_depend(JOB_DEPEND_TYPE_ON, pattr);
```
Comment (vstumpf) [inline src/server/req_register.c:359]: 1 extra indent here

File: src/server/req_register.c [hunk 6]
```diff
@@ -327,23 +356,23 @@ req_register(struct batch_request *preq)
...
 						if (pdep == 0) {
 							/* on "on", see if child already registered */
 							revtype = type ^ (JOB_DEPEND_TYPE_BEFORESTART -
 								JOB_DEPEND_TYPE_AFTERSTART);
 							pdep = find_depend(revtype, pattr);
 							if (pdep == 0) {
 								/* no "on" and no prior - return error */
-								rc = PBSE_BADDEPEND;
+							rc = PBSE_BADDEPEND;
```
Comment (vstumpf) [inline src/server/req_register.c:367]: one indent too less

File: src/server/req_register.c [hunk 7]
```diff
@@ -327,23 +356,23 @@ req_register(struct batch_request *preq)
...
 							} else {
 								pdj = find_dependjob(pdep,
 									preq->rq_ind.rq_register.rq_child);
 								if (pdj) {
 									/* has prior register, update it */
 									(void)strcpy(pdj->dc_svr,
 										preq->rq_ind.rq_register.rq_svr);
-								}
+									}
```
Comment (vstumpf) [inline src/server/req_register.c:375]: 1 indent too much

File: src/server/req_register.c [hunk 8]
```diff
@@ -531,18 +559,24 @@ post_doq(struct work_task *pwt)
 
 	if (preq->rq_reply.brp_code) {
 		/* request was rejected */
+		if (preq->rq_reply.brp_code == PBSE_HISTJOBID)
+			log_eventf(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, jobid,
+				   "%s %s, dependency satisfied", preq->rq_ind.rq_register.rq_parent, msg_historyjobid);
+		else if (preq->rq_reply.brp_code == PBSE_HISTDEPEND)
+			log_eventf(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, jobid,
+				   "%s %s", preq->rq_ind.rq_register.rq_parent, msg_histdepend);
+		else {
```
Comment (vstumpf) [inline src/server/req_register.c:568]: you don't need the braces around the else case now.

**Comment by arungrover** [comment]
I've addressed your comments @vstumpf

**Comment by vstumpf** [comment]
SmokeTest.test_round_robin failed on the sanitize build, so I restarted it.

**Comment by arungrover** [comment]
Thanks @vstumpf

**Pull request merged**